File: Code/Java_Code/Chapter04_Kinematics/BeanBag.java
import javax.swing.*;
import java.awt.*;
import javax.swing.border.BevelBorder;
import java.awt.event.*;
import java.util.Random;
import javax.swing.Timer;

public class BeanBag extends JFrame implements ActionListener
{
  private JTextField vxTextField;
  private JTextField vzTextField;

  private JLabel vxLabel;
  private JLabel vzLabel;

  private JButton fireButton;
  private JButton resetButton;
  private JPanel drawingPanel;
  private GridBagConstraints gbc;

  private double z;     //  altitude of beanbag
  private double z0;    //  initial altitude of beanbag
  private double vz0;   //  initial vertical velocity
  private double x;     //  horizontal location
  private double x0;    //  initial horizontal location
  private double vx0;   //  initial horizontal velocity
  private double time;      

  //  These elements are used to control the execution
  //  speed of the game. Without them, the game would
  //  run too quickly.
  private GameUpdater gameUpdater;
  private Timer gameTimer;

  //  These fields are for the images used in the game.
  private ImageIcon playerIcon;
  private ImageIcon backgroundIcon;

  public BeanBag() {

    //  Initialize the beanbag parameters.
    z = 1.7;
    z0 = 1.7;
    vz0 = 0.0;
    x = 0.5;
    x0 = 0.5;
    vx0 = 0.0;
    time = 0.0;

    //  Create a Timer object that will be used
    //  to slow the action down and an ActionListener
    //  that the Timer will call. The timeDelay variable
    //  is the time delay in milliseconds.
    gameUpdater = new GameUpdater();
    int timeDelay = 50;
    gameTimer = new Timer(timeDelay, gameUpdater);

    //  Set up images
    playerIcon = new ImageIcon("BeanbagTosser.jpg");
    backgroundIcon = new ImageIcon("Background.jpg");

    //  Create JTextField objects to display the outcome.
    vxTextField = new JTextField("2.5",8);
    vzTextField = new JTextField("2.0",8);

    //  Create some JLabels
    vxLabel = new JLabel("Initial horizontal velocity, m/s");
    vzLabel = new JLabel("Initial vertical velocity, m/s");

    //  Create a JButton that will start the box moving
    fireButton = new JButton("Fire");
    fireButton.setBorder(new BevelBorder(BevelBorder.RAISED));
    fireButton.setPreferredSize(new Dimension(60,35));
    fireButton.addActionListener(this);

    //  Create a JButton that will update the drawing area.
    resetButton = new JButton("Reset");
    resetButton.setBorder(new BevelBorder(BevelBorder.RAISED));
    resetButton.setPreferredSize(new Dimension(60,35));
    resetButton.addActionListener( new ActionListener() {
      public void actionPerformed(ActionEvent ae) {
        //  stop the timer.
        gameTimer.stop();

        //  Reset the box and ball location and time;
        z = 1.7;
        z0 = 1.7;
        vz0 = 0.0;
        x = 0.5;
        x0 = 0.5;
        vx0 = 0.0;
        time = 0.0;

        //  Update the display.
        updateDisplay();
      }  
    });

    //  Create a JTextArea that will display the results
    drawingPanel = new JPanel();
    drawingPanel.setPreferredSize(new Dimension(251, 151));

    //  Place components on a panel using a GridBagLayout
    JPanel westPanel = new JPanel();
    GridBagLayout gridBagLayout1 = new GridBagLayout();
    westPanel.setLayout(gridBagLayout1);

    int col;
    int row;
    int numCol = 1;
    int numRow = 1;
    Insets insets = new Insets(5, 3, 5, 3);

    row = 0;
    col = 0;
    gbc = new GridBagConstraints(col, row, numCol, numRow,
                 0.0, 0.0, GridBagConstraints.EAST,
                 GridBagConstraints.NONE, insets, 0, 0);
    gridBagLayout1.setConstraints(vxLabel, gbc);

    col = 1;
    gbc = new GridBagConstraints(col, row, numCol, numRow,
                 0.0, 0.0, GridBagConstraints.WEST,
                 GridBagConstraints.NONE, insets, 0, 0);
    gridBagLayout1.setConstraints(vxTextField, gbc);

    row = 1;
    col = 0;
    gbc = new GridBagConstraints(col, row, numCol, numRow,
                 0.0, 0.0, GridBagConstraints.EAST,
                 GridBagConstraints.NONE, insets, 0, 0);
    gridBagLayout1.setConstraints(vzLabel, gbc);

    col = 1;
    gbc = new GridBagConstraints(col, row, numCol, numRow,
                 0.0, 0.0, GridBagConstraints.WEST,
                 GridBagConstraints.NONE, insets, 0, 0);
    gridBagLayout1.setConstraints(vzTextField, gbc);

    row = 2;
    col = 0;
    gbc = new GridBagConstraints(col, row, numCol, numRow,
                 0.0, 0.0, GridBagConstraints.EAST,
                 GridBagConstraints.NONE, insets, 0, 0);
    gridBagLayout1.setConstraints(fireButton, gbc);

    row = 3;
    col = 0;
    gbc = new GridBagConstraints(col, row, numCol, numRow,
                 0.0, 0.0, GridBagConstraints.EAST,
                 GridBagConstraints.NONE, insets, 0, 0);
    gridBagLayout1.setConstraints(resetButton, gbc);

    westPanel.add(vxLabel);
    westPanel.add(vxTextField);
    westPanel.add(vzLabel);
    westPanel.add(vzTextField);
    westPanel.add(fireButton);
    westPanel.add(resetButton);

    //  The drawing panel.
    JPanel eastPanel = new JPanel();
    GridBagLayout gridBagLayout2 = new GridBagLayout();
    eastPanel.setLayout(gridBagLayout2);

    row = 0;
    col = 0;
    gbc = new GridBagConstraints(col, row, numCol, numRow,
                 0.0, 0.0, GridBagConstraints.CENTER,
                 GridBagConstraints.NONE, 
                 new Insets(10, 10, 10, 20), 0, 0);
    gridBagLayout2.setConstraints(drawingPanel, gbc);

    eastPanel.add(drawingPanel);

    //  Add the JPanel objects to the content pane
    getContentPane().setLayout(new BorderLayout());
    getContentPane().add(westPanel, BorderLayout.WEST);
    getContentPane().add(eastPanel, BorderLayout.EAST);

    //  Add a title to the JFrame, size it, and make it visible.
    setTitle("Beanbag Game");
    setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
    setBounds(100,100,600,300);
    setVisible(true);

    //  Update the GUI display
    updateDisplay(); 
  }

  //  The actionPerformed() method is called when 
  //  the "Start" button is pressed. 
  public void actionPerformed(ActionEvent event) {
    
    //  Get the initial velocities from the textfield
    vx0 = Double.parseDouble(vxTextField.getText());
    vz0 = Double.parseDouble(vzTextField.getText());

    //  Start the box sliding using a Timer object
    //  to slow down the action.
    gameTimer.start();
  }

  //  This method redraws the GUI display.
  private void updateDisplay() {
    Graphics g = drawingPanel.getGraphics();
    int width = drawingPanel.getWidth() - 1;
    int height = drawingPanel.getHeight() - 1;

    g.clearRect(0, 0, width, height);
    g.drawRect(0, 0, width, height); 

    //  Draw the beanbag tosser and background.
    g.drawImage(backgroundIcon.getImage(), 1, 1, 
                  width-2, 123, drawingPanel);
    int zLocation = 125 - 67;
    g.drawImage(playerIcon.getImage(), 7, zLocation, 
                  51, 67, drawingPanel);

    g.setColor(Color.BLACK);
    g.drawLine(0, 125, width, 125);
    g.drawLine(150, 125, 150, height);
    g.drawLine(175, 125, 175, height);
    g.drawLine(200, 125, 200, height);
    g.drawLine(225, 125, 225, height);
    g.drawString("10", 155, 143);
    g.drawString("20", 180, 143);
    g.drawString("50", 205, 143);
    g.drawString("0", 230, 143);

    //  Update the position of the beanbag
    //  on the screen.
    int xPosition = (int)(100.0*x);
    double deltaZ = z - 1.25;
    int zPosition = (int)(125 - 100.0*deltaZ);
    g.fillOval(xPosition, zPosition, 10, 10);

  }

  public static void main(String args[]) {
    BeanBag gui = new BeanBag();
  }

  //  This ActionListener is called by the Timer
  class GameUpdater implements ActionListener {
    public void actionPerformed(ActionEvent event) {

      //  Update the time and compute the new position
      //  of the beanbag. 
      double timeIncrement = 0.05;
      time += timeIncrement; 

      //  There is no force in the x-direction, so the
      //  new x location is the initial x location plus
      //  the product of the horizontal velocity and time.
      x = x0 + vx0*time;

      //  The z-location is influenced by the acceleration
      //  due to gravity.
      double g = -9.81;
      z = z0 + vz0*time + 0.5*g*time*time;

      //  Update the display
      updateDisplay();

      //  If the beanbag hits the ground, stop 
      //  the simulation.
      if ( z <= 1.4 ) {
        gameTimer.stop();
      }
    }
  }
}

File: Code/Csharp_Code/Chapter04_Kinematics/BeanBag.cs
using System;
using System.Windows.Forms;
using System.Drawing;

public class BeanBag : Form
{
  private Label vxLabel;
  private Label vzLabel;

  private TextBox vxTextBox;
  private TextBox vzTextBox;

  private Button fireButton;
  private Button resetButton;

  private Panel drawingPanel;

  //  This field is for the images used in the game.
  private Image playerIcon;

  //  The Timer is used to control the execution speed
  //  of the game.
  private Timer gameTimer;

  private double z;     //  altitude of beanbag
  private double z0;    //  initial altitude of beanbag
  private double vz0;   //  initial vertical velocity
  private double x;     //  horizontal location
  private double x0;    //  initial horizontal location
  private double vx0;   //  initial horizontal velocity
  private double time;      

  public BeanBag() {

    //  Initialize the beanbag parameters.
    z = 1.7;
    z0 = 1.7;
    vz0 = 0.0;
    x = 0.5;
    x0 = 0.5;
    vx0 = 0.0;
    time = 0.0;

    //  Set up images
    playerIcon = Image.FromFile("BeanbagTosser.jpg");

    //  Create a Timer object that will be used
    //  to slow the action down.
    gameTimer = new Timer();
    gameTimer.Interval = 50;  //  delay in milliseconds.
    gameTimer.Tick += new EventHandler(ActionPerformed);

    //  Create some Labels
    vxLabel = new Label();
    vxLabel.Text = "Initial horizontal velocity, m/s";
    vxLabel.Font = new Font(vxLabel.Font, FontStyle.Bold);
    vxLabel.Top = 30;
    vxLabel.Left = 10;
    vxLabel.Width = 180;

    vzLabel = new Label();
    vzLabel.Text = "Initial vertical velocity, m/s";
    vzLabel.Font = new Font(vzLabel.Font, FontStyle.Bold);
    vzLabel.Top = 60;
    vzLabel.Left = 10;
    vzLabel.Width = 180;

    //  Create TextBox objects to display the outcome.
    vxTextBox = new TextBox();
    vxTextBox.Width = 50;
    vxTextBox.Text = "2.5";
    vxTextBox.AutoSize = true;
    vxTextBox.Top = vxLabel.Top;
    vxTextBox.Left = 200;

    vzTextBox = new TextBox();
    vzTextBox.Width = 50;
    vzTextBox.Text = "2.0";
    vzTextBox.AutoSize = true;
    vzTextBox.Top = vzLabel.Top;
    vzTextBox.Left = 200;

    //  Create Button objects 
    int buttonHeight = 30;
    int buttonWidth = 50;
    int buttonLeft = 50;

    fireButton = new Button();
    fireButton.Text = "Fire";
    fireButton.Height = buttonHeight;
    fireButton.Width = buttonWidth;
    fireButton.Top = 100;
    fireButton.Left = buttonLeft;
    fireButton.Click += new EventHandler(FireButtonClicked);

    resetButton = new Button();
    resetButton.Text = "Reset";
    resetButton.Height = buttonHeight;
    resetButton.Width = buttonWidth;
    resetButton.Top = 160;
    resetButton.Left = buttonLeft;
    resetButton.Click += new EventHandler(ResetButtonClicked);

    //  Create a drawing panel.
    drawingPanel = new Panel();
    drawingPanel.Width = 251;
    drawingPanel.Height = 151;
    drawingPanel.Left = 280;
    drawingPanel.Top = 50;
    drawingPanel.BorderStyle = BorderStyle.FixedSingle;

    //  Add the GUI components to the Form
    this.Controls.Add(vxLabel);
    this.Controls.Add(vzLabel);
    this.Controls.Add(vxTextBox);
    this.Controls.Add(vzTextBox);
    this.Controls.Add(fireButton);
    this.Controls.Add(resetButton);
    this.Controls.Add(drawingPanel);

    // Set the size and title of the form
    this.Height = 300;
    this.Width = 600;
    this.Text = "Beanbag Game";

    //  Center the form on the screen and make
    //  it visible.
    this.StartPosition = FormStartPosition.CenterScreen;
    this.Visible = true;

    //  Update the GUI display
    UpdateDisplay();
  }

  //  Event handling method for the "Fire" button
  public void FireButtonClicked(object source, EventArgs e) {

    //  Extract initial data from the textfields.
    vx0 = Convert.ToDouble(vxTextBox.Text);
    vz0 = Convert.ToDouble(vzTextBox.Text);

    //  Start the box sliding using a Timer object
    //  to slow down the action.
    gameTimer.Start();
  }

  //  Event handling method for the "Reset" button
  public void ResetButtonClicked(object source, EventArgs e) {
    //  stop the timer.
    gameTimer.Stop();

    //  Reset the box and ball location and time;
    z = 1.7;
    z0 = 1.7;
    vz0 = 0.0;
    x = 0.5;
    x0 = 0.5;
    vx0 = 0.0;
    time = 0.0;

    //  Update the display.
    UpdateDisplay();
  }

  //  This method redraws the GUI display.
  private void UpdateDisplay() {
    Graphics g = drawingPanel.CreateGraphics();
    int width = drawingPanel.Width - 1;
    int height = drawingPanel.Height - 1;

    //  Clear the current display.
    g.Clear(drawingPanel.BackColor);

    //  Draw the beanbag tosser.
    int zLocation = 125 - 67;
    g.DrawImage(playerIcon, 7, zLocation, 51, 67);

    Pen blackPen = new Pen(Color.Black, 2);
    g.DrawLine(blackPen, 0, 125, width, 125);
    g.DrawLine(blackPen, 150, 125, 150, height);
    g.DrawLine(blackPen, 175, 125, 175, height);
    g.DrawLine(blackPen, 200, 125, 200, height);
    g.DrawLine(blackPen, 225, 125, 225, height);

    SolidBrush brush = new SolidBrush(Color.Black);
    Font font = new Font("Arial", 12);
    g.DrawString("10", font, brush, 152, 127);
    g.DrawString("20", font, brush, 177, 127);
    g.DrawString("50", font, brush, 202, 127);
    g.DrawString("0", font, brush, 227, 127);

    //  Update the position of the beanbag
    //  on the screen.
    int xPosition = (int)(100.0*x);
    double deltaZ = z - 1.25;
    int zPosition = (int)(125 - 100.0*deltaZ);
    g.FillEllipse(brush, xPosition, zPosition, 10, 10);

    //  Clean up the Graphics object.
    g.Dispose();
  }

  //  This method is called by the Timer every 0.05 seconds.
  public void ActionPerformed(object source, EventArgs e) {
    //  Update the time and compute the new position
    //  of the beanbag. 
    double timeIncrement = 0.05;
    time += timeIncrement; 

    //  There is no force in the x-direction, so the
    //  new x location is the initial x location plus
    //  the product of the horizontal velocity and time.
    x = x0 + vx0*time;

    //  The z-location is influenced by the acceleration
    //  due to gravity.
    double g = -9.81;
    z = z0 + vz0*time + 0.5*g*time*time;

    //  Update the display
    UpdateDisplay();

    //  If the beanbag hits the ground, stop 
    //  the simulation.
    if ( z <= 1.4 ) {
      gameTimer.Stop();
    }
  }

  static void Main() {
    Application.Run(new BeanBag());
  }
}
